bigtop-tests/smoke-tests/logstash/TestLogstashSmoke.groovy

```diff
@@ -0,0 +1,71 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import org.junit.BeforeClass
+import org.junit.AfterClass
+import org.apache.bigtop.itest.shell.Shell
+import static org.junit.Assert.assertTrue
+import static org.junit.Assert.assertFalse
+import org.junit.Test
+import org.apache.bigtop.itest.TestUtils
+import org.junit.runner.RunWith
+import static org.junit.Assert.assertEquals
+
+class TestLogstashSmoke {
+  static Shell sh = new Shell("/bin/bash -s");
+
+  static final String LOGSTASH_CMD = "/usr/lib/logstash/bin/logstash "
+  static final String TINYBENCH_CONFIG = "resources/generator.conf "
+  static final String FILEPIPELINE_CONFIG = "resources/filepipeline.conf "
+
+  @AfterClass
+  public static void LogstashCleanUp() {
+    sh.exec("rm -f resources/output.log");
+    assertTrue("Logstash cleanup failed. ", sh.getRet() == 0);
+  }
+
+  @Test
+  public void LogstashTinyBenchTest() {
+    /* Tiny benchmark with logstash Generator-plugin */
+    sh.exec("time "
+      + LOGSTASH_CMD
+      + "-f "
+      + TINYBENCH_CONFIG
+    );
+    def out = sh.out.join('\n');
+    assertFalse("Logstash tiny benchmark failed", (out =~ /ERROR/).find());
+  }
+
+  @Test
+  public void LogstashFilePipelineTest() {
```

Comment:
This test assumes /bigtop-home is the directory into which Bigtop is copied. It's always true for the docker provisioner, but I think it depends on the user when deploying a real cluster.
In addition, this test failed with the following error in my environment, for some reason (it looks like output file was not generated, but I didn't investigate in detail):

```
TestLogstashSmoke > LogstashFilePipelineTest FAILED
    java.lang.AssertionError: Logstash output did not match expected output. expected:<0> but was:<2>
```

How about removing LogstashFilePipelineTest for now (and adding it in the future if needed), because LogstashTinyBenchTest seems to be enough as a minimal smoke test if it also checks the status code of the logstash command?

Reply:
Yes, it makes sense. I'll update it soon, thanks

Reply:
Thanks for the response Yuqi, one additional comment. I tried the latest PR on Ubuntu 16.04, and got the following error.

```
$ ./gradlew toolchain
$ ./gradlew clean logstash-pkg

(snip)

+ rake bootstrap
mkdir -p vendor
mkdir vendor/_
mkdir -p build
mkdir -p build/bootstrap
Installing minitar 0.5.4 because the build process needs it.
[bootstrap] Fetching and installing gem: minitar (0.5.4)
Successfully installed minitar-0.5.4
1 gem installed
Installing bundler ~> 1.9.4 because the build process needs it.
[bootstrap] Fetching and installing gem: bundler (~> 1.9.4)
Successfully installed bundler-1.9.10
1 gem installed
Invoking bundler install...
Error Bundler::InstallError, retrying 1/10
An error occurred while installing childprocess (4.0.0), and Bundler cannot continue.
Make sure that `gem install childprocess -v '4.0.0'` succeeds before bundling.

(snip)

Too many retries, aborting, caused by Bundler::InstallError
rake aborted!
Bundler::InstallError: An error occurred while installing childprocess (4.0.0), and Bundler cannot continue.
Make sure that `gem install childprocess -v '4.0.0'` succeeds before bundling.
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/installer.rb:145:in `install_gem_from_spec'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/installer.rb:269:in `block in install_sequentially'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/installer.rb:268:in `install_sequentially'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/installer.rb:90:in `run'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/installer.rb:18:in `install'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/cli/install.rb:80:in `run'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/cli.rb:155:in `install'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/cli.rb:10:in `start'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/lib/bootstrap/bundler.rb:160:in `execute_bundler'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/lib/bootstrap/bundler.rb:131:in `block in execute_bundler_with_retry'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/lib/bootstrap/bundler.rb:129:in `block in execute_bundler_with_retry'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/lib/bootstrap/bundler.rb:203:in `capture_stdout'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/lib/bootstrap/bundler.rb:128:in `execute_bundler_with_retry'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/lib/bootstrap/bundler.rb:118:in `invoke!'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/rakelib/vendor.rake:122:in `block in (root)'
Gem::RuntimeRequirementNotMetError: childprocess requires Ruby version >= 2.4.0. The current ruby version is 2.3.0.
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/source/rubygems.rb:116:in `block in install'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/rubygems_integration.rb:131:in `preserve_paths'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/source/rubygems.rb:113:in `install'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/installer.rb:107:in `install_gem_from_spec'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/installer.rb:269:in `block in install_sequentially'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/installer.rb:268:in `install_sequentially'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/installer.rb:90:in `run'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/installer.rb:18:in `install'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/cli/install.rb:80:in `run'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/cli.rb:155:in `install'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/vendor/bundle/jruby/2.3.0/gems/bundler-1.9.10/lib/bundler/cli.rb:10:in `start'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/lib/bootstrap/bundler.rb:160:in `execute_bundler'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/lib/bootstrap/bundler.rb:131:in `block in execute_bundler_with_retry'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/lib/bootstrap/bundler.rb:129:in `block in execute_bundler_with_retry'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/lib/bootstrap/bundler.rb:203:in `capture_stdout'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/lib/bootstrap/bundler.rb:128:in `execute_bundler_with_retry'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/lib/bootstrap/bundler.rb:118:in `invoke!'
/home/vagrant/bigtop/output/logstash/logstash-5.4.1/rakelib/vendor.rake:122:in `block in (root)'
Tasks: TOP => bootstrap => vendor:all => vendor:gems
(See full trace by running task with --trace)
debian/rules:31: recipe for target 'override_dh_auto_build' failed
make[1]: *** [override_dh_auto_build] Error 1
make[1]: Leaving directory '/home/vagrant/bigtop/output/logstash/logstash-5.4.1'
debian/rules:28: recipe for target 'build' failed
make: *** [build] Error 2
dpkg-buildpackage: error: debian/rules build gave error exit status 2
debuild: fatal error at line 1376:
dpkg-buildpackage -rfakeroot -D -us -uc -b
 failed

> Task :logstash-deb FAILED
```

The message says the ruby version is 2.3.0, but 2.4.0 seems to be enabled via RVM.

```
$ ruby -v
ruby 2.4.0p0 (2016-12-24 revision 57164) [x86_64-linux]
$ rvm list
=* ruby-2.4.0 [ x86_64 ]

# => - current
# =* - current && default
#  * - default
```

Am I missing something?

Reply:
The PR was updated for removing LogstashFilePipelineTest and using `$LOGSTASH_HOME `instead of a hard-coded path.

For ubuntu-16.04 build issue, let me look into it on my environment and will give you feddback later, thanks.

Reply:
Sorry for the late response.
The root cause is: 
Childprocess-4.0.0 is newly released in **June 18, 2020**. 
But it is not compatible with 5.4.1.
Specify childprocess to 3.0.0.
Thanks!